src/corona_hakab_model/matrix_analysis/MatrixAnalysis.py

```diff
@@ -0,0 +1,135 @@
+        Path(self.EXPORT_MATRIX_DIR).mkdir(parents=True, exist_ok=True)
+
+    def export_raw_matrices_to_csv(self):
+        """
+        run over all types of connections and export each of them to a csv file
+        """
+        for connection_type in range(self.matrix_data.depth):
+            self.export_matrix_to_csv(connection_type)
+        # all connections combined
+        self.export_matrix_to_csv()
+
+    def export_matrix_to_csv(self, connection_type=None):
+        if connection_type is None:
+            connection_matrix = sum(self.matrix_data.matrix)
+            connection_type = "all"
+        else:
+            connection_matrix = self.matrix_data.matrix[connection_type]
+
+        # get the indices and values of all non-zero values in the matrix
+        (connection, agent), value = connection_matrix.nonzero(), connection_matrix.data
+        connections = map(lambda x, y, z: [x, y, z], agent, connection, value)
+        connections = sorted(connections)
+
+        headers = ['agent', 'connection', 'value']
+        export_file_path = self.EXPORT_MATRIX_DIR + self.EXPORT_MATRIX_FILE.format(connection_type)
+        with open(export_file_path, 'w', newline='') as matrix_file:
+            csv_writer = csv.writer(matrix_file, delimiter=',')
+            csv_writer.writerows([headers]+connections)
+
+    def analyze_histograms(self):
+        """
+        run over all types of connections and analyze the connection histogram
+        """
+        for connection_type in range(self.matrix_data.depth):
+            self.analyze_connection_histogram(connection_type)
+        self.analyze_connection_histogram()
+
+    def analyze_connection_histogram(self, connection_type=None):
+        if connection_type is None:
+            connection_matrix = sum(self.matrix_data.matrix)
+            connection_type = "all"
+        else:
+            connection_matrix = self.matrix_data.matrix[connection_type]
+
+        # get the number of connections for each agent, i.e. the number of non-zero rows for each column
+        distribution_of_connections = connection_matrix.getnnz(0)
+
+        average_connection = np.mean(distribution_of_connections)
+        median_connection = np.median(distribution_of_connections)
+        max_connection = np.max(distribution_of_connections)
+        min_connection = np.min(distribution_of_connections)
+        connections_histogram, _ = np.histogram(distribution_of_connections,
+                                                bins=(max_connection-min_connection+1),
+                                                range=(min_connection - 0.5, max_connection + 0.5))
+        bins = np.arange(min_connection, max_connection+1)
+
+        self.export_histogram_to_csv(connections_histogram, bins,
+                                     average_connection, median_connection, connection_type)
+        self.plot_histogram(connections_histogram, bins, average_connection, median_connection, connection_type)
```

Comment:
Is seems like you bundle here things that shouldnt be bundled. Is sounds like this function just needs to return the statistics and then when calling export_hist.. it should receive histogram and save it.
Same goes for the plot

Reply:
קיבלתי את זה שצריך להפריד בין התפקידים של הפונקציות אבל לדעתי אין סיבה לתת ליוזר להחליט אם לייצא את הקבצים או לשמור את התמונות כי זאת כל המטרה של המודול... הוא מודול נפרד והוא אמור לרוץ רק כשרוצים לייצר את הדברים האלה, אז אין סיבה להעמיס על היוזר עם אופציות שונות להרצה